src/bank00.asm

```diff
@@ -998,78 +993,73 @@ drawMetaTile:
-    ld   B, $00                                        ;; 00:05a3 $06 $00
-    add  HL, BC                                        ;; 00:05a5 $09
-    ld   C, [HL]                                       ;; 00:05a6 $4e
-    pop  HL                                            ;; 00:05a7 $e1
-    push BC                                            ;; 00:05a8 $c5
-    ld   A, [HL+]                                      ;; 00:05a9 $2a
-    ld   HL, wBackgroundGraphicsTileMapping            ;; 00:05aa $21 $70 $d0
-    ld   C, A                                          ;; 00:05ad $4f
-    ld   B, $00                                        ;; 00:05ae $06 $00
-    add  HL, BC                                        ;; 00:05b0 $09
-    ld   L, [HL]                                       ;; 00:05b1 $6e
-    pop  BC                                            ;; 00:05b2 $c1
-    ld   H, C                                          ;; 00:05b3 $61
-    call storeDEatBackgroundDrawPosition               ;; 00:05b4 $cd $95 $04
+    ld   A, [HL+]
+    ld   B, A
+    ld   A, [HL+]
+    ld   C, A
+    ld   A, [HL+]
+    ld   D, $00
+    ld   E, [HL]
+    push BC
+    ld   HL, wBackgroundGraphicsTileMapping
+    add  HL, DE
+    ld   C, [HL]
+    ld   HL, wBackgroundGraphicsTileMapping
+    ld   E, A
+    add  HL, DE
+    ld   B, [HL]
+    pop  DE
+    push BC
+    ld   A, D
+    ld   D, $00
+    ld   HL, wBackgroundGraphicsTileMapping
+    add  HL, DE
+    ld   C, [HL]
+    ld   HL, wBackgroundGraphicsTileMapping
+    ld   E, A
+    add  HL, DE
+    ld   B, [HL]
+    pop  DE
+    pop  HL
+    call storeBCDEinBGM
     call popBankNrAndSwitch                            ;; 00:05b7 $cd $0a $2a
     ret                                                ;; 00:05ba $c9
 
-; Exactly the same as getTileInfoPointer
-getTileInfoPointer2:
-    ld   L, A                                          ;; 00:05bb $6f
-    ld   H, $00                                        ;; 00:05bc $26 $00
-    ld   D, H                                          ;; 00:05be $54
-    ld   E, L                                          ;; 00:05bf $5d
-    add  HL, HL                                        ;; 00:05c0 $29
-    add  HL, DE                                        ;; 00:05c1 $19
-    add  HL, HL                                        ;; 00:05c2 $29
-    ld   A, [wTileDataTablePointer.High]               ;; 00:05c3 $fa $93 $d3
-    ld   D, A                                          ;; 00:05c6 $57
-    ld   A, [wTileDataTablePointer]                    ;; 00:05c7 $fa $92 $d3
-    ld   E, A                                          ;; 00:05ca $5f
-    add  HL, DE                                        ;; 00:05cb $19
-    ret                                                ;; 00:05cc $c9
+; Store BCDE at metatile with upper left corner at HL.
+; Waits until PPU mode 0 (HBlank) or 1 (VBlank).
+; Return: a = d, bc = $001f, de = de, hl = hl + 4
+storeBCDEinBGM:
```

Comment:
1. I think I'd prefer `storeBCDEinVRAM` as a name. BGM is too well established as BackGround Music. Plus, there is nothing constraining this to only be used for backgrounds.
2. This looks pretty tight in timing, but it looks like it'll always make it under the wire. `storeDEinVRAM` has the comment "mode 2 (OAM scan) has enough time to finish". OAM scan is 80 dots, which should mean 20 cycles. `storeDEinVRAM` looks like 10 cycles total starting with the `and STATF_BUSY` until the writes are finished. This looks like 20 cycles starting counting in the same place. That's a lot harder to be confident about without testing.

One cycle *could* be trimmed by replacing:
`ld bc, $001f`
`add hl, bc`
... with:
`ld a, l`
`add $1f`
`ld l, a`
... which is valid for metatiles, but only for metatiles. (In that case the name should be changed to something like `writeMetatileToVRAM`.)

Once again, this is *probably* fine and if it's not then just play testing will *probably* show visible glitches. I'm just wondering if my math seems correct and whether you've thought through this.

Reply:
My reasoning: 

Worst case scenario the ldh a, [rSTAT] just misses by a dot the start of hblank, so it loops again before exiting. That's 15 cycles or 60 dots. I think I have a way to reduce this worst case to 9 cycles that I'll commit after testing. 

After exiting the loop there are 15 cycles (another 60 dots) that need to have access to VRAM. So 120 dots total needed (worst case) as currently written.

According to pandocs, hblank is 87 dots minimum and OAM scan (the next mode after hblank) is always 80 dots. Both allow writing to VRAM, so we've got 167 dots to work with. Plenty of room for the call. 

![Screenshot_20240729-222707.png](https://github.com/user-attachments/assets/fa709f33-aa6e-450a-b35a-e6929d19ea21)

Reply:
I like writeMetatileToVRAM, and your L only optimization. I'll change those.

Reply:
Wouldn't the worst-case be that the `ldh a, [rSTAT]` just catches the *end* of hblank and runs 100% in OAM scan?

`storeDEinVRAM` used to use a double-loop to make sure it started at the *beginning* of hblank but I optimized it since it had plenty of time to run in OAM scan.

Reply:
Thinking about it more, this could be changed to use `set 5, l` to bring it down from five to just two cycles.

Reply:
I think this is now resolved